### PR TITLE
actionnetwork improvements

### PR DESCRIPTION
**Pull request by liewegas**
- sync reminders
- use per-action forms
- pass referrer data

### CODE REVIEW

**Comment by liewegas** [comment]
https://actionnetwork.org/docs/v2/record_submission_helper references one other field that is not included here:
```
referrer | string |   | The referrer code of this submission. Equivalent to someone taking action with the url argument ?referrer=[your referrer code]. Must be a valid Action Network referrer code. Read-only. Corresponds to the referrers chart in this action's manage page. Action Network-only.
```
I'm not sure if it's applicable here?

File: app/integration/actionnetwork.py [hunk 1]
```diff
@@ -23,13 +25,59 @@ class ActionNetworkError(Exception):
...
+            an_id, va_action = get_form_ids(response.json()["identifiers"])
+            if an_id and va_action:
+                r[va_action] = an_id
+    return r
+
+
+def post_person(info, action, api_key):
     from common.apm import tracer
 
+    forms = setup_action_forms()
```
Comment (nickcatal) [inline app/integration/actionnetwork.py:72]: We can definitely cache this (for a very very long time) in redis using django's cache framework

File: app/integration/actionnetwork.py [hunk 2]
```diff
@@ -23,13 +25,59 @@ class ActionNetworkError(Exception):
...
+def get_form_ids(ids):
+    an_id = None
+    va_action = None
+    for i in ids:
+        (a, b) = i.split(":")
+        if a == "action_network":
+            an_id = b
+        elif a == "voteamerica":
+            va_action = b
+    return an_id, va_action
```
Comment (nickcatal) [inline app/integration/actionnetwork.py:37]: can you use vars that are longer than 1 character. this is a bit hard to follow 😄

File: app/integration/actionnetwork.py [hunk 3]
```diff
@@ -23,13 +25,59 @@ class ActionNetworkError(Exception):
...
+    return r
+
+
+def post_person(info, action, api_key):
     from common.apm import tracer
 
+    forms = setup_action_forms()
+    form_id = forms.get(action)
+
     with tracer.trace("an.person.add", service="actionnetwork"):
```
Comment (nickcatal) [inline app/integration/actionnetwork.py:75]: we're no longer using the `add` endpoint, so likely want to change this

File: app/integration/actionnetwork.py [hunk 4]
```diff
@@ -98,20 +145,29 @@ def _sync_item(item, subscriber_id):
...
+                        "postal_code": item.zipcode,
+                        "country": "US",
+                    },
+                ],
+            },
+            "action_network:referrer_data": {
+                "source": item.source,
+                "website": item.embed_url,
+                "email_referrer": item.email_referrer,
+                "mobile_referrer": item.mobile_referrer,
```
Comment (nickcatal) [inline app/integration/actionnetwork.py:167]: thanks for this. this is going to be super useful

File: app/integration/actionnetwork.py [hunk 5]
```diff
@@ -23,13 +25,59 @@ class ActionNetworkError(Exception):
...
+    for action in ["lookup", "register", "ballotrequest", "reminderrequest"]:
+        if action not in r:
+            # This code runs once per action, ever.
+            logger.info(f"Creating action form for {action}")
+            response = requests.post(
+                FORM_ENDPOINT,
+                headers={"OSDI-API-Token": settings.ACTIONNETWORK_KEY},
+                json={
+                    "identifiers": [f"voteamerica:{action}"],
+                    "title": action,
```
Comment (nickcatal) [inline app/integration/actionnetwork.py:59]: We definitely want to prefix the names of these forms with "VoteAmerica" or something so that subscribers know that our tools are the origin of the form when they do targeting

File: app/integration/actionnetwork.py [hunk 6]
```diff
@@ -23,13 +25,59 @@ class ActionNetworkError(Exception):
...
+    for form in response.json()["_embedded"]["osdi:forms"]:
+        an_id, va_action = get_form_ids(form["identifiers"])
+        if an_id and va_action:
+            r[va_action] = an_id
+
+    for action in ["lookup", "register", "ballotrequest", "reminderrequest"]:
+        if action not in r:
+            # This code runs once per action, ever.
+            logger.info(f"Creating action form for {action}")
+            response = requests.post(
```
Comment (nickcatal) [inline app/integration/actionnetwork.py:54]: def want to wrap this (and the GET above) inside an APM tracer

**Comment by liewegas** [comment]
So the good news is that we can push all the improved referral info to actionnetwork by just blowing away the ``integration_link`` table and repeating the full sync.  The bad/annoying news is that means turning off the ``ACTIONNETWORK_DAILY_SYNC`` temporarily while the full sync runs manually, and then turning it back on afterwards (next day or whatever).

File: app/integration/actionnetwork.py [hunk 7]
```diff
@@ -3,33 +3,99 @@
...
 
-ADD_ENDPOINT = "https://actionnetwork.org/api/v2/people/"
+FORM_ENDPOINT = "https://actionnetwork.org/api/v2/forms/"
+ADD_ENDPOINT = "https://actionnetwork.org/api/v2/forms/{form_id}/submissions/"
 
 logger = logging.getLogger("integration")
 
+ACTIONS = ["lookup", "register", "ballotrequest", "reminderrequest"]
+
+CACHE_KEY = "lookup", "register", "ballotrequest", "reminderrequest"
```
Comment (nickcatal) [inline app/integration/actionnetwork.py:26]: Wouldn't we need separate forms and cache keys for each subscriber?

File: app/integration/actionnetwork.py [hunk 8]
```diff
@@ -3,33 +3,99 @@
...
+                        "identifiers": [f"voteamerica:{action}"],
+                        "title": f"VoteAmerica {action}",
+                        "origin_system": "voteamerica",
+                    },
+                )
+                an_id, va_action = get_form_ids(response.json()["identifiers"])
+                if an_id and va_action:
+                    forms[va_action] = an_id
+
+    cache.set(CACHE_KEY, forms)
```
Comment (nickcatal) [inline app/integration/actionnetwork.py:82]: By default django only caches for 5 minutes https://docs.djangoproject.com/en/3.0/ref/settings/#timeout

We likely want to set this timeout to be something much longer (and allow it to be adjusted in settings.) I'd imagine at a minimum we'd want it to be cached for a day?

File: app/integration/actionnetwork.py [hunk 9]
```diff
@@ -3,33 +3,99 @@
...
+        if action not in forms:
+            # This code runs once per action, ever.
+            logger.info(f"Creating action form for {action}")
+            with tracer.trace("an.form.create", service="actionnetwork"):
+                response = requests.post(
+                    FORM_ENDPOINT,
+                    headers={"OSDI-API-Token": settings.ACTIONNETWORK_KEY},
+                    json={
+                        "identifiers": [f"voteamerica:{action}"],
+                        "title": f"VoteAmerica {action}",
```
Comment (nickcatal) [inline app/integration/actionnetwork.py:74]: This is random, but this will render to `VoteAmerica registration`, since we're lowercasing our action names. Can we make it `VoteAmerica BallotRequest` or `VoteAmerica Registration` (the full camelcase class name)

File: app/integration/actionnetwork.py [hunk 10]
```diff
@@ -98,20 +162,29 @@ def _sync_item(item, subscriber_id):
...
+                        "address_lines": [item.address1, item.address2],
+                        "locality": item.city,
+                        "region": item.state_id,
+                        "postal_code": item.zipcode,
+                        "country": "US",
+                    },
+                ],
+            },
+            "action_network:referrer_data": {
+                "source": item.source,
```
Comment (nickcatal) [inline app/integration/actionnetwork.py:181]: Does it make sense to have a default source, maybe `voteamerica_{actionname}` or something? Just thinking, I dunno here.

File: app/integration/actionnetwork.py [hunk 11]
```diff
@@ -3,33 +3,99 @@
...
+
+    forms = {}
+    with tracer.trace("an.form", service="actionnetwork"):
+        response = requests.get(
+            FORM_ENDPOINT, headers={"OSDI-API-Token": settings.ACTIONNETWORK_KEY},
+        )
+        for form in response.json()["_embedded"]["osdi:forms"]:
+            an_id, va_action = get_form_ids(form["identifiers"])
+            if an_id and va_action:
+                forms[va_action] = an_id
```
Comment (nickcatal) [inline app/integration/actionnetwork.py:62]: What happens if a subscriber has hundreds of forms and the ActionNetwork response paginates?
